--- gcc-9.3.0-raw/gcc/sanitizer.def	2020-03-12 07:07:21.000000000 -0400
+++ gcc-9.3.0/gcc/sanitizer.def	2020-11-20 02:47:12.151216993 -0500
@@ -583,5 +583,11 @@
 		      "__sanitizer_cov_trace_switch",
 		      BT_FN_VOID_UINT64_PTR, ATTR_NOTHROW_LEAF_LIST)
 
+/* Object coverage */
+DEF_SANITIZER_BUILTIN(BUILT_IN_SANITIZER_OBJ_COV_TRACE_PC,
+		      "__sanitizer_obj_cov_trace_pc",
+		      BT_FN_VOID, ATTR_NOTHROW_LEAF_LIST)
+
+
 /* This has to come after all the sanitizer builtins.  */
 DEF_BUILTIN_STUB(END_SANITIZER_BUILTINS, (const char *)0)
